proc/cut.go

```diff
@@ -1,74 +1,109 @@
 package proc
 
 import (
+	"errors"
 	"fmt"
+	"strings"
 
+	"github.com/mccanne/zq/pkg/zeek"
 	"github.com/mccanne/zq/pkg/zson"
+	"github.com/mccanne/zq/pkg/zval"
 )
 
+var ErrNoField = errors.New("cut field not found")
+
 type Cut struct {
 	Base
-	fields     []string
-	sawRecord  bool
-	seenFields uint64
+	fields   []string
+	cutmap   map[int]*zson.Descriptor
+	nblocked int
 }
 
 func NewCut(c *Context, parent Proc, fields []string) *Cut {
-	return &Cut{Base: Base{Context: c, Parent: parent}, fields: fields}
+	return &Cut{
+		Base:   Base{Context: c, Parent: parent},
+		fields: fields,
+		cutmap: make(map[int]*zson.Descriptor),
+	}
 }
 
-// Check the bitmap of fields that we've seen.  If a requested field
-// never appeared in the input, emit a warning about it.  This should be
-// called once when this proc reaches the end of the stream.
-func (c *Cut) WarnUnseen() {
-	if !c.sawRecord {
-		return
+func (c *Cut) lookup(in *zson.Descriptor) *zson.Descriptor {
+	d, ok := c.cutmap[in.ID]
+	if ok {
+		return d
 	}
-	for i, name := range c.fields {
-		if (c.seenFields & (1 << i)) == 0 {
-			c.Warnings <- fmt.Sprintf("Cut field %s not present in input", name)
+	var columns []zeek.Column
+	for _, field := range c.fields {
+		colno, ok := in.ColumnOfField(field)
+		if !ok {
+			// a field is missing... block this descriptor
+			c.cutmap[in.ID] = nil
+			c.nblocked++
+			return nil
+		}
+		columns = append(columns, in.Type.Columns[colno])
+	}
+	out := c.Resolver.GetByColumns(columns)
+	c.cutmap[in.ID] = out
+	return out
+}
+
+// CreateCut returns a new record value derived by keeping only the fields
+// specified by name in the fields slice.
+func (c *Cut) cut(d *zson.Descriptor, in *zson.Record) (*zson.Record, error) {
+	var zv zval.Encoding
+	for _, column := range d.Type.Columns {
+		colno, ok := in.ColumnOfField(column.Name)
+		if !ok {
+			return nil, ErrNoField
 		}
+		zv = zval.Append(zv, in.Slice(colno), zeek.IsContainerType(column.Type))
 	}
+	return zson.NewRecordNoTs(d, zv), nil
+}
+
+func (c *Cut) warn() {
+	if len(c.cutmap) > c.nblocked {
+		return
+	}
```

Comment:
Feels like it should only warn if nothing was emitted... maybe just do `if len(c.cutmap) > 0 { return }`

Reply:
This is what the check does.  I wrote it your way first and that has a bug :-)  The cutmap tracks both descriptors that match and those that don't see you need to check that the number in the table is greater than the number that didn't match.